test/functional/plugin/lsp_spec.lua

```diff
@@ -3623,4 +3623,103 @@ describe('LSP', function()
       eq(expected, result)
     end)
   end)
+
+  describe('vim.lsp._watchfiles', function()
+    it('should send workspace/didChangeWatchedFiles when watched files change', function()
+      local expected_handlers = {
+        {NIL, {}, {method="shutdown", client_id=1}};
+        {NIL, {}, {method="reregister", client_id=1}};
+        {NIL, {}, {method="register", client_id=1}};
+        {NIL, nil, {method="get_root", bufnr=1, client_id=1}};
+      }
+      local client
+      local root_dir
+      test_rpc_server {
+        test_name = "basic_watchfiles";
+        on_init = function(_client, _)
+          client = _client
+          client.request('get_root')
+        end;
+        on_exit = function(code, signal)
+          eq(0, code, "exit code")
+          eq(0, signal, "exit signal")
+        end;
+        on_handler = function(err, result, ctx)
+          if ctx.method == 'get_root' then
+            root_dir = result
+            result = nil -- don't make an assertion on this value
+          elseif ctx.method == 'register' then
+            exec_lua([[
+            local root_dir = ...
+            vim.lsp.handlers["client/registerCapability"](NIL, {
+              registrations = {
+                {
+                  id = "basic_watchfiles_0",
+                  method = "workspace/didChangeWatchedFiles",
+                  registerOptions = {
+                    watchers = {
+                      {
+                        globPattern = {
+                          baseUri = vim.uri_from_fname(root_dir),
+                          pattern = "**/*.watch",
+                        },
+                        kind = 7,
+                      },
+                    },
+                  },
+                },
+              },
+            }, {client_id=1})
+            ]], root_dir)
+            write_file(root_dir..'/test.nowatch', '')
+            os.remove(root_dir..'/test.nowatch')
+            write_file(root_dir..'/test.watch', '')
+            write_file(root_dir..'/test.watch', 'some change')
+            os.remove(root_dir..'/test.watch')
```

Comment:
This sequence of events runs into the race condition I described above where a "created" event followed by a "deleted" event might end up being interpreted as a "deleted" event. This makes the test very flaky for me locally. Any suggestions to make this test more deterministic?

Reply:
The new iteration of this test I just pushed waits for the "created" and "changed" events to come through before doing the delete, which seems to fix this at the expense of waiting for an extra 100ms timeout between the event being noticed and the notification being sent.